Add Humidifier support to zwave_js

EiNSTeiN- opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a `ZWaveHumidifier` class which implements a `HumidifierEntity` for z-wave devices supporting HumidityControl CCs.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- HumidityControl CCs added to zwave_js in https://github.com/zwave-js/node-zwave-js/pull/4048
- Python constants added in https://github.com/home-assistant-libs/zwave-js-server-python/pull/341

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/21758

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

EiNSTeiN- committed:
Add Z-Wave Humidifier support

EiNSTeiN- commented on homeassistant/components/humidifier/__init__.py:
```diff
@@ -169,6 +172,9 @@ def state_attributes(self) -> dict[str, Any]:
             data[ATTR_MODE] = self.mode
 
+        if supported_features & SUPPORT_OPERATING_STATE:
+            data[ATTR_OPERATING_STATE] = self.operating_state
```
ZWave humidifiers expose the current operating state of the humidifier, which is different from the "mode" in which the humidifier is set. For instance, in "auto" mode, the operating state will go back and forth between `Humidifying` and `Idle`. I added this new attribute since I think this exposes useful information that had no good way of being exposed before. My intention is to follow up with UX changes to display this attribute in the Humidifier card when available, just like the Thermostat card does.

raman325 replied:
Changes to the humidifier platform need to be a separate PR. You can either mark this one as draft and wait until the platform changes are merged and rebase, or you can remove the support for this feature in this PR and submit it as a second one, your call.

EiNSTeiN- replied:
Opened: https://github.com/home-assistant/core/pull/65855

MartinHjelmare replied:
I suggest removing all features from this PR that we won't accept immediately to not block this PR.

EiNSTeiN- commented on homeassistant/components/zwave_js/discovery.py:
```diff
@@ -492,6 +495,17 @@ def get_config_parameter_discovery_schema(
+            property={HUMIDITY_CONTROL_MODE_PROPERTY},
+            type={"number"},
+        ),
+        allow_multi=True,
```
The same z-wave node will likely have a humidifier and a thermostat, which is why this discovery entry is placed above the `climate` ones, and this one has `allow_multi`. Is this the correct way of doing this?

raman325 replied:
`allow_multi` is only needed when the same ZWave Value is used to create multiple entities. In this case I don't think it is needed.

homeassistant commented:
Hello @EiNSTeiN-,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`2f5e949e120db53178a1305c8306a71412027118`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;65847&#x2F;commits&#x2F;2f5e949e120db53178a1305c8306a71412027118) - This commit has something that looks like an email address (fc@francoischagnon.net). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

raman325 commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,188 @@
+            if HumidityControlMode.HUMIDIFY in self._current_mode.metadata.states:
+                self._attr_device_class = HumidifierDeviceClass.HUMIDIFIER
+            else:
+                self._attr_device_class = HumidifierDeviceClass.DEHUMIDIFIER
```
maybe submit a PR for the platform to have a new device class that means both?

EiNSTeiN- replied:
Opened https://github.com/home-assistant/core/pull/65858

EiNSTeiN- replied:
To avoid blocking this PR I'd make the assumption that we can set the device class to humidifier even though the device does both?

EiNSTeiN- replied:
After reading through https://github.com/home-assistant/architecture/issues/310 I found https://github.com/home-assistant/architecture/issues/310#issuecomment-719752004 which seems to point towards splitting the humidifying and de-humidifying capabilities into two entities. This would mean the z-wave `Humidity control mode` would need to map to the on/off state of each device in this way:

| Humidifier entity | De-humidifier entity | Z-Wave Humidity Control Mode |
|----|----|----|
| OFF | OFF | OFF |
| ON | OFF | Humidify |
| OFF | ON | De-humidify |
| ON | ON | Auto |

Thoughts on this approach? Still seems weird to me because two entities need to know about each other's state to control the underlying z-wave value correctly.

MartinHjelmare replied:
I'd suggest to simplify and do last service call wins. Ie, not have any interdependence between entities.

EiNSTeiN- replied:
I'll move this back to draft and make the changes to split these in two entities

EiNSTeiN- replied:
I'm thinking about this a bit more, there's no need for two entities to have a dependency on each other. This can be mediated through the underlying zwave value which would be shared between both entities regardless.

On the humidifier side we will have:
```python
def turn_on():
  if current_mode == OFF:
    current_mode = HUMIDIFY
  elif current_mode == DEHUMIDIFY:  # turn on humidifier from "de-humidify"-only by setting the mode to auto
    current_mode = AUTO

def turn_off():
  if current_mode = AUTO: # turn off humidifier from "auto" by setting the mode to de-humidify only
    current_mode = DEHUMIDIFY
  elif current_mode = HUMIDIFY:
    current_mode = OFF
```

On the de-humidifier side:
```python
def turn_on():
  if current_mode == OFF:
    current_mode = DEHUMIDIFY
  elif current_mode == HUMIDIFY: # turn on de-humidifier from "humidify"-only by setting the mode to auto
    current_mode = AUTO

def turn_off():
  if current_mode = AUTO: # turn off de-humidifier from "auto" by setting the mode to humidify only
    current_mode = HUMIDIFY
  elif current_mode = DEHUMIDIFY:
    current_mode = OFF
```

raman325 reviewed (changes requested):
see responses to comments

EiNSTeiN- commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,253 @@
+        attrs = {}
+
+        if state := self.operating_state:
+            attrs[ATTR_OPERATING_STATE] = state
```
I updated this PR to use `extra_state_attributes` for this `operating_state` attribute to avoid blocking on a decision for https://github.com/home-assistant/core/pull/65855

MartinHjelmare replied:
I suggest we leave this out of the implementation for now. It's better to wait until we know how we want to express this generally.

EiNSTeiN- replied:
done

EiNSTeiN- committed:
Remove operating_state attribute

EiNSTeiN- commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,225 @@
+            ):
+                new_mode = HumidityControlMode.HUMIDIFY
+            else:
+                new_mode = HumidityControlMode.DEHUMIDIFY
```
Would it be better to keep track of the last mode that was active before turning off and go back to that mode when turning on? Is there a way to persist that information across reboot or is that not necessary?

MartinHjelmare replied:
We should receive the current state from the driver when starting the integration. Is that not the case?

MartinHjelmare replied:
I understand now. If the device is off, we don't know what mode it was before that.

MartinHjelmare replied:
I don't think we need to make the last not off state persistent across restarts. We can't guarantee that the user didn't change modes during the time Home Assistant was shutdown. So loading a stored state from file may be incorrect for the user anyway.

It's ok to store a last non off state in memory of we think that's the best user experience.

EiNSTeiN- replied:
The device is "on" when the state is anything other than "off". Turning the device "off" involves updating the value of `self._current_mode`, and inversely when turning the device back on, we need to set the mode to something other than Off. My question is what should that new mode be (choices are: auto, humidify, de-humidify). Should I keep track of the previous mode and go back to it instead of the logic I implemented here? If the previous mode is stored in an instance variable on the entity, then if I were to set the device to "off" and restart home-assistant, which mode should we go back to then?

EiNSTeiN- replied:
>  If the device is off, we don't know what mode it was before that.

Yes this is correct, sorry github didn't show your comment while I was writing mine...

EiNSTeiN- replied:
After splitting this in two entities, there is no longer a "mode" exposed on the humidifier, only a on/off state, so this is no longer necessary

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,225 @@
+        """Set new mode."""
+
+        if self._current_mode is None:
+            return None
```
Nothing is checking this return value.
```suggestion
            return
```

EiNSTeiN- replied:
can't apply this suggestion:

![image](https://user-images.githubusercontent.com/2053039/152703561-55dbd3f7-2c18-4741-9464-b9bb3236a8d1.png)

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,225 @@
+            if self._setpoint_value and self._setpoint_value.metadata.min:
+                min_value = self._setpoint_value.metadata.min
+        # In case of any error, we fallback to the default
+        except (IndexError, ValueError, TypeError):
```
When can these exceptions be raised?

EiNSTeiN- replied:
This was copied from `climate.py` but I think this was added because of the temperature conversion, it may not be necessary here. I removed it.

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,225 @@
+    @property
+    def mode(self) -> str | None:
+        """Return the current humidity control mode."""
+        if self._current_mode:
```
Generally it's nicer to use guard clause pattern and return the short fast answer first. Ie invert the check and return None if true like we do above.

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,225 @@
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Set new target humidity."""
+        if self._current_mode:
```
Please invert the check and return early if true.

EiNSTeiN- committed:
Split humidifier and de-humidifier in two entities

EiNSTeiN- commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,214 @@
+        super().__init__(config_entry, client, info)
+
+        self._attr_name = f"{self._attr_name} {self.device_class}"
+        self._attr_unique_id = f"{self._attr_unique_id}_{self.device_class}"
```
this turned out to be necessary, otherwise both the humidifier and the de-humidifier will get the same unique_id and name, and only the first entity will be accessible

MartinHjelmare replied:
Don't use the device class. Define a string that the integration is in control of and use that as suffix.

EiNSTeiN- commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,214 @@
+        else:
+            return
+
+        await self.info.node.async_set_value(self._current_mode, new_mode)
```
turn_on and turn_off implement this logic: https://github.com/home-assistant/core/pull/65847#discussion_r800217658

EiNSTeiN- commented:
Updated with two entities, one for humidifier & one for de-humidifier.

EiNSTeiN- committed:
Use self.info.primary_value, and move duplicate code to base class

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,182 @@
+        return max_value
+
+
+class ZWaveHumidifier(ZWaveBaseHumidifier):
```
Please use entity descriptions instead to describe the entity and use the same entity class for both entity kinds. See eg the binary sensor platform.

EiNSTeiN- replied:
Updated!

Shulyaka commented on homeassistant/components/zwave_js/discovery.py:
```diff
@@ -492,6 +495,16 @@ def get_config_parameter_discovery_schema(
         ),
     ),
+    # humidifier
+    # thermostats supporting mode (and optional setpoint)
```
```suggestion
    # hygrostats supporting mode (and optional setpoint)
```

Shulyaka commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,182 @@
+"""Representation of Z-Wave thermostats."""
```
```suggestion
"""Representation of Z-Wave humidifiers."""
```

Shulyaka commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,182 @@
+        client: ZwaveClient,
+        info: ZwaveDiscoveryInfo,
+    ) -> None:
+        """Initialize thermostat."""
```
```suggestion
        """Initialize humidifier."""
```

Shulyaka commented on tests/components/zwave_js/test_humidifier.py:
```diff
@@ -0,0 +1,951 @@
+
+
+async def test_humidifier(hass, client, climate_adc_t3000, integration):
+    """Test a thermostat v2 command class entity."""
```
```suggestion
    """Test a humidifier v2 command class entity."""
```
Not sure what v2 means here

Shulyaka commented on tests/components/zwave_js/test_humidifier.py:
```diff
@@ -0,0 +1,951 @@
+
+
+async def test_dehumidifier(hass, client, climate_adc_t3000, integration):
+    """Test a thermostat v2 command class entity."""
```
```suggestion
    """Test a dehumidifier v2 command class entity."""
```

Shulyaka commented on tests/components/zwave_js/test_humidifier.py:
```diff
@@ -0,0 +1,951 @@
+"""Test the Z-Wave JS climate platform."""
```
```suggestion
"""Test the Z-Wave JS humidifier platform."""
```

Shulyaka reviewed (changes requested):
Not sure that all my fixes are correct but there are a number of references to thermostats and climate platform that might be confusing.
Otherwise, would be really nice to have the Z-Wave support!

EiNSTeiN- committed:
Fix references to thermostat in comments

EiNSTeiN- committed:
Use entity description for humidifier and dehumidifier

EiNSTeiN- commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -0,0 +1,219 @@
+        self.entity_description = description
+
+        self._attr_name = f"{self._attr_name} {description.name_suffix}"
+        self._attr_unique_id = f"{self._attr_unique_id}_{description.name_suffix}"
```
I added a `name_suffix` as part of the entity description instead of the device class

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -28,6 +33,43 @@
+    on_mode=HumidityControlMode.HUMIDIFY,
+    inverse_mode=HumidityControlMode.DEHUMIDIFY,
+    setpoint_type=HumidityControlSetpointType.HUMIDIFIER,
+    name_suffix="humidifier",
```
We normally use the key attribute for this purpose.

MartinHjelmare commented on homeassistant/components/zwave_js/humidifier.py:
```diff
@@ -28,6 +33,43 @@
+DEHUMIDIFIER_ENTITY_DESCRIPTION = ZwaveHumidifierEntityDescription(
+    key="1",
+    device_class=HumidifierDeviceClass.DEHUMIDIFIER,
+    on_mode=HumidityControlMode.DEHUMIDIFY,
```
Are some of the custom attributes required? In that case it's better to create a mixin dataclass for the required attributes so they can be typed correctly as required instead of optional.

EiNSTeiN- replied:
Updated this as well, sorry I haven't done any typed python before, I was not familiar with dataclasses.

EiNSTeiN- committed:
Use mixin class for required keys

EiNSTeiN- committed:
Remove unneeded guard clause for required parameter

EiNSTeiN- commented on tests/components/zwave_js/test_humidifier.py:
```diff
@@ -0,0 +1,951 @@
+            "label": "Humidity control mode",
+            "states": {"0": "Off", "1": "Humidify", "2": "De-humidify", "3": "Auto"},
+        },
+        "value": int(HumidityControlMode.HUMIDIFY),
```
I'm unsure why this test is failing, it is passing when I run it locally... I'm calling `node.receive_event(event)` above which should set the value of that z-wave value to `HUMIDIFY`, this is what the `vlaue` attribute on this line should be set to. Is there a better way to mock the state of a z-wave value for these tests?

raman325 replied:
not sure why it works locally but you need to await hass.async_block_till_done after receiving the event so that the state change can be processed

EiNSTeiN- replied:
Added

raman325 replied:
sorry I misread, I thought you were trying to process a state change in HA, but the issue is the node's state.

Out of curiosity, locally are you using the pip installed latest version of `zwave-js-server-python` or are you using a local version of `master`?

EiNSTeiN- replied:
I see this is also part of the diff
```
E         +  'newValue': 1,
E         +  'prevValue': 0,
```
which matches keys that are in the `event` above:
```
                "newValue": int(HumidityControlMode.HUMIDIFY),
                "prevValue": int(HumidityControlMode.OFF),
```
but locally when I run the test these keys are not even set in the dict being compared on L186...

EiNSTeiN- replied:
> Out of curiosity, locally are you using the pip installed latest version of zwave-js-server-python or are you using a local version of master?

Still using a local version since I started this branch before my changes were merged... I might need to rebase to pick up a version that has the humidity control changes?

EiNSTeiN- committed:
Fix tests against correct version of zwave_js_server_python

EiNSTeiN- committed:
Add fixture and test for a humidifier missing the corresponding setpoint

EiNSTeiN- commented:
I added coverage to make the codecov CI check happier, in the process I had to add a new fixture with one of the `Humidity Control Setpoint` removed.

I'm not sure how this would happen in reality, perhaps a weird device might use the same setpoint for both humidifier and de-humidifier, or advertise one of the control modes even though it doesn't actually support changing the corresponding setpoint. I'm not aware this is an issue with any device today but I guess we will find out as this gets rolled out.

Since there are guard clauses whenever `self._setpoint` is referenced, and its type is defined as `_setpoint: ZwaveValue | None = None`, I removed the `raise ValueError` line I had added in case `self._setpoint` is None in the constructor. Should I add some sort of warning if this is the case? Or disable the entity? What would be a sane default behavior?

EiNSTeiN- commented:
@MartinHjelmare @raman325 anything else I can do to move this PR forward?

MartinHjelmare reviewed (approved):
Looks good! One line seems to be missing test coverage. Otherwise good to go.

EiNSTeiN- committed:
Add test for a humidifier that does not support de-humidify mode

EiNSTeiN- commented:
> Looks good! One line seems to be missing test coverage. Otherwise good to go.

Added a test for it!

MartinHjelmare commented on tests/components/zwave_js/conftest.py:
```diff
@@ -276,6 +276,28 @@ def climate_radio_thermostat_ct100_plus_different_endpoints_state_fixture():
+
+
+@pytest.fixture(name="climate_adc_t3000_missing_mode_state", scope="session")
+def climate_adc_t3000_missing_mode_state_fixture():
```
Instead of adding a fixture json file for each case, maybe it's easy enough to modify the fixture data for a good device and remove the data that should be missing? Each time we load a new fixture file it takes a significant relative amount of time in the test run.

EiNSTeiN- replied:
Done

MartinHjelmare commented:
There are test failures.

EiNSTeiN- committed:
Fix tests again against zwave_js_server_python 0.35.1, remove unnecessary fixtures json

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Please link a docs PR where we add the new platform support to the ha-platforms front matter at the top of the zwave_js page.

EiNSTeiN- commented:
> Please link a docs PR where we add the new platform support to the ha_platforms front matter at the top of the zwave_js page.

Added: https://github.com/home-assistant/home-assistant.io/pull/21758
